### PR TITLE
Update flux typings

### PR DESCRIPTION
**Pull request by tht13**
Based off of PR #9534 
I have updated the flux typings to improve generics in the different Store classes.

case 2. Improvement to existing type definition.
- documentation or source code reference which provides context for the suggested changes.
  - it has been reviewed by a DefinitelyTyped member.
### Container Class

Before:

``` typescript
class Container {
    static create(base: React.ComponentClass<any>, options?: any): React.ComponentClass<any>;
}
```

After:

``` typescript
interface RealOptions {
    pure?: boolean;
    withProps?: boolean;
}
class Container {
    static create<TComponent>(base: React.ComponentClass<TComponent>, options?: RealOptions): React.ComponentClass<TComponent>;
}
```
### Store Class

Before:

``` typescript
class Store {
    constructor(dispatcher: Flux.Dispatcher<any>);
    addListener(callback: Function): { remove: Function };
    getDispatcher(): Flux.Dispatcher<any>;
    __onDispatch(payload: any): void;
}
```

After:

``` typescript
class Store<TPayload> {
    constructor(dispatcher: Flux.Dispatcher<TPayload>);
    addListener(callback: Function): fbEmitter.EventSubscription;
    getDispatcher(): Flux.Dispatcher<TPayload>;
    __onDispatch(payload: TPayload): void;
}
```
### ReduceStore Class

Before:

``` typescript
class ReduceStore<T> extends Store {
    reduce(state: T, action: any): T;
}
```

After:

``` typescript
class ReduceStore<T, TPayload> extends Store<TPayload> {
    reduce(state: T, action: TPayload): T;
}
```
### MapStore Class

Before:

``` typescript
class MapStore<K extends string | number, V, TPayload> extends ReduceStore<immutable.Map<K, V>, TPayload> {
        getAll(keys: any, prev?: any): any;
}
```

After:

``` typescript
class MapStore<K extends string | number, V> extends ReduceStore<any> {
        getAll(keys: immutable.Iterable.Indexed<K>, prev?: immutable.Map<K, V>): immutable.Map<K, V>;
}
```

### CODE REVIEW

**Comment by DovydasNavickas** [comment]
As you are introducing a breaking change for `Store` class (adding non-optional generic type parameter), you should provide an upgrade guide for that.

An example can be found here: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/8422

File: fbemitter/fbemitter.d.ts [hunk 1]
```diff
@@ -64,4 +64,8 @@ declare module 'fbemitter' {
          
     }
     
+}
+
+declare module 'fbemitter' {
+    export = FBEmitter;
 }
```
Comment (GiedriusGrabauskas) [inline fbemitter/fbemitter.d.ts]: Missing newline at end of file

File: flux/flux.d.ts [hunk 2]
```diff
@@ -71,21 +73,43 @@ declare module "flux" {
...
+    
     export class Container {
         constructor();
         /**
         * Create is used to transform a react class into a container
         * that updates its state when relevant stores change.
         * The provided base class must have static methods getStores() and calculateState().
         */
-        static create(base: React.ComponentClass<any>, options?: any): React.ComponentClass<any>;
+        static create<P, V extends {}>(base: React.ComponentClass<P>, options?: V & RealOptions): React.ComponentClass<P>;
```
Comment (GiedriusGrabauskas) [inline flux/flux.d.ts]: Rename to `TComponent`

File: fbemitter/fbemitter.d.ts (hunk 1 above)
Comment (GiedriusGrabauskas) [inline fbemitter/fbemitter.d.ts]: Missing newline at end of file

File: flux/flux.d.ts (hunk 2 above)
Comment (GiedriusGrabauskas) [inline flux/flux.d.ts]: `Options` have only `pure` and `withProps` available variables. Why you added generic `V`?

File: flux/flux.d.ts [hunk 3]
```diff
@@ -71,21 +73,43 @@ declare module "flux" {
...
     import React = __React;
+    import fbEmitter = FBEmitter;
+    import immutable = Immutable;
+    
+    /**
+     * Default options to create a Container with
+     * 
+     * @interface RealOptions
+     */
+    interface RealOptions {
```
Comment (GiedriusGrabauskas) [inline flux/flux.d.ts]: All properties are optional. So, you should at each optional property add a `?` at the end of the property name.

**Comment by vvakame** [comment]
👀

**Comment by GiedriusGrabauskas** [comment]
```ts
export class Container {
 // ...
static create<TComponent, V extends {}>(base: React.ComponentClass<TComponent>, options?: V & RealOptions): React.ComponentClass<TComponent>;
 // ...
}
```

Options have only `pure` and `withProps` available properties. Why did you add generic type `V` in `create` function?

**Source code:**  [facebook/flux/FluxContainer.js#L25](https://github.com/facebook/flux/blob/dbe8d5ea15dd89dc584562f3c03096cd12453856/src/container/FluxContainer.js#L25)

**Comment by tht13** [comment]
Cannot remember why I did that, have reverted it.

**Comment by tht13** [comment]
Updated PR message to include breaking changes

**Comment by DovydasNavickas** [comment]
:+1: 
Good job! :clap:

**Comment by GiedriusGrabauskas** [comment]
Cool :wink: :+1:

**Pull request merged**